wasm/wat/lexer.go

```diff
@@ -0,0 +1,239 @@
+			if s == 0 {
+				return fmt.Errorf("%d:%d found an invalid byte in block comment: 0x%x", line, col, b1)
+			}
+			p = p + s - 1 // -1 as the for loop will + 1: This optimizes speed of tokenization over block comments.
+			continue
+		}
+
+		// One design-affecting constraint is that all tokens begin and end with a 7-bit ASCII character. This
+		// simplifies line and column counting and how to detect the end of a token.
+		//
+		// Note: Even though string allows unicode contents it is enclosed by ASCII double-quotes (").
+		//
+		// See https://www.w3.org/TR/wasm-core-1/#characters%E2%91%A0
+
+		// While many tokens can be single character only parens must be: Handle these more cheaply.
+		switch b1 {
+		case '(':
+			if e := parser(source, tokenLParen, line, col, p, p+1); e != nil {
+				return e
+			}
+			continue
+		case ')':
+			if e := parser(source, tokenRParen, line, col, p, p+1); e != nil {
+				return e
+			}
+			continue
+		}
+
+		// Track positions passed to the parser
+		// beginLine == line because no token is allowed to include an unescaped '\n'
+		beginCol := col  // the start column of the token (fixed)
+		beginPos := p    // the start position of the token (fixed)
+		peekPos := p + 1 // when finished scanning, this becomes endPos (the position after the token).
+
+		tok := firstTokenByte[b1]
+		switch tok {
+		case tokenSN: // min 2 bytes for sign and number; ambiguous: could be tokenFN
+			return fmt.Errorf("%d:%d TODO: signed", line, col)
+		case tokenUN: // min 1 byte; ambiguous when >=3 bytes as could be tokenFN
+			if peekPos < length {
+				peeked := source[peekPos]
+				if peeked == 'x' {
+					return fmt.Errorf("%d:%d TODO: hex", line, col)
+				}
+			loop:
+				// Start after the number and run until the end. Note all allowed characters are single byte.
+				for ; peekPos < length; peekPos = peekPos + 1 {
+					switch source[peekPos] {
+					case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '_':
+						p = peekPos
+						col = col + 1
+					default:
+						break loop // end of this token (or malformed, which the next loop will notice)
+					}
+				}
+			}
+		case tokenString: // min 2 bytes for empty string ("")
+			hitQuote := false
+			// Start at the second character and run until the end. Note UTF-8 (multi-byte) characters are allowed.
+			for peekPos < length {
+				peeked := source[peekPos]
+				if peeked == '"' { // TODO: escaping and banning disallowed characters like newlines.
+					hitQuote = true
+					break
+				}
+
+				col = col + 1
+				s := utf8Size[peeked] // While unlikely, it is possible the current byte is invalid unicode
+				if s == 0 {
+					return fmt.Errorf("%d:%d found an invalid byte in string token: 0x%x", line, col, peeked)
+				}
+				peekPos = peekPos + s
+			}
+
+			if !hitQuote {
+				return fmt.Errorf("%d:%d expected end quote", line, col)
+			}
+
+			// set the position to after the quote
+			p = peekPos
```

Comment:
As the `peekPos` is *on* the quote :D
```suggestion
			p = peekPos
			// set the position to after the quote
```